### ISSUE TITLE
Unit Testing Custom Filter

### ISSUE THREAD

**Issue by steven-rothwell** [microsoft/FeatureManagement-Dotnet#231]
I'm trying to unit test my custom filter.  I have a current question about this on [StackOverflow](https://stackoverflow.com/questions/76165742/how-to-mock-featurefilterevaluationcontext-parameters).

Immediately, I try to get the settings for the filter.
```
public virtual Task<Boolean> EvaluateAsync(FeatureFilterEvaluationContext context)
{
    _settings = context.Parameters.Get<IdFilterSettings>();
```

Where IdFilterSettings looks like this:

```
internal class IdFilterSettings
{
    public String[]? AllowedIds { get; set; }
    public String[]? BlockedIds { get; set; }
}
```

When calling it through an actual API call, _settings properly gets set to a new instance of IdFilterSettings with AllowedIds and BlockedIds properly populated.

I try to unit test this and use fake data in FeatureFilterEvaluationContext with this code:

```
var allowedIds = new string[2] { "someId1", "someId2" };
var blockedIds = new string[1] { "someId3" };
var parameters = new Dictionary<string, string?>
{
    { "AllowedIds", JsonSerializer.Serialize(allowedIds) },
    { "BlockedIds", JsonSerializer.Serialize(blockedIds) }
};
var context = new FeatureFilterEvaluationContext
{
    Parameters = new ConfigurationBuilder().AddInMemoryCollection(parameters).Build()
};

_filter = new CustomIdFilter(_httpContextAccessor.Object);

var result = await _filter.EvaluateAsync(context);
```

If I put a break point after _settings = context.Parameters.Get<IdFilterSettings>();, settings is a new instance of IdFilterSettings, but AllowedIds and BlockedIds are both null.

I tried Mocking FeatureFilterEvaluationContext, but the Parameters property is not virtual, so I cannot Mock it.  I tried Mocking IConfiguration and setting it to Parameters, but IConfiguration.Get<IdFilterSettings> is not allowed as it is also not virtual.  How can we unit test our custom filters?

**Comment by jimmyca15** [comment]
Hey @steven-rothwell thanks for reaching out. The issue here should be that the configuration dictionary you're building is not quite what .NET Core makes when it reads from app settings.

The dictionary that you pass in to AddInMemoryCollection should actually look like this, according to your scenario.

```
{
    { "AllowedIds:0", "someId1" },
    { "AllowedIds:1", "someId2" }
    { "blockedIds:0", "someId1" }
}
```

The idea here is that arrays are flattened. You can see I've appended the index of each element in the array to the configuration setting key.

Let me know if this helps !

**Comment by rossgrambo** [comment]
I wanted to callout that in [2.6.0-preview](https://github.com/microsoft/FeatureManagement-Dotnet/releases/tag/2.6.0-preview) we added an `Exclusion` list on TargetingFilters, where you can explicitly block specific users or groups from the filter. Not sure if it's relevant for you, but it sounded similar to your example custom filter so I figured I'd mention it.

**Comment by steven-rothwell** [comment]
@jimmyca15, thank you so much for getting back to me so quickly!  I spent about 6 hours trying to figure this out today.  Your suggestion worked!  Truly appreciate it!

@rossgrambo, that's cool!  We needed a little bit more customization.  I ended up customizing both the FeatureGate and FeatureManager to allow the coder to tell the custom filter where to look for the id.  (Ex: query param, claims, or given value)  But, thanks for letting me know!
